### DISCUSSION TITLE
make TypeORM work with SvelteKit

### DISCUSSION THREAD

**Discussion by Ennoriel** [sveltejs/kit#10376] [Q&A] [answered]
I'm trying to make TypeORM work with SvelteKit and following the Getting Started page of their documentation (https://typeorm.io/) but encontered some difficulties to make it work: the js anotation (@Entity) added with the library `reflect-metadata` does not seem to be recognized well: 

first error: 
```
Column type for User2#name is not defined and cannot be guessed. Make sure you have turned on an "emitDecoratorMetadata": true option in tsconfig.json. Also make sure you have imported "reflect-metadata" on top of the main entry file in your application (before any entity imported).If you are using JavaScript instead of TypeScript you must explicitly provide a column type.
```

second error (when explicitely specifying a type):
```
EntityMetadataNotFoundError: No metadata for "User" was found.
```

What I tried:
- set the lib as dependencies (and not dev dep): "reflect-metadata", "typeorm", "pg"
- set the lib in `ssr.external` in vite config

There is already a discussion but with a beta version and thought it would be better to open a fresh one: https://github.com/sveltejs/kit/discussions/3334

Anyone figured out how to make it work or willing to help me? :)

Thanks ;)

**Comment by ngramp** [comment, marked as answer]
First of all ensure you have emitDecoratorMetadata set in `tsconfig.json` as per the docs:
```jsonc
 "compilerOptions": {
    ...    
    // needed for typeORM
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true
  },`
``` 
And include reflect-metadata in `vite.config.js` as you mentioned:
```javascript
export default defineConfig({
  ...
  ssr: {
    external: ['reflect-metadata'],
  },
});
``` 
Despite all this effort with metadata reflection, there have been previous bugs with it, so ensure that every column in your entities has a labelled `type`:
```typescript
@Entity({ name: 'test' })
export class Test extends BaseEntity {
  ...
  @Column({ type: 'varchar'})
  name!: string | null;
}
``` 
I personally have had some issues with `DataSource` being re instantiated, so I put it inside of a singleton in `$lib/db.ts`:
```typescript
import { Test } from './entities/Test';

class TypeOrm {
  private static instance: Promise<DataSource> | null = null;

  private constructor() {
    // Private constructor to prevent external instantiation
  }

  public static getDb(): Promise<DataSource> {
    if (!TypeOrm.instance) {
      TypeOrm.instance = new DataSource({
        type: 'postgres',
        host: 'localhost',
        port: 5432,
        username: 'svelte', // make these .env
        password: 'sveltepassword',
        database: 'sveltetypeorm',
        synchronize: true,
        entities: [Test],
        migrations: [],
        subscribers: [],
        logging: true,
      })
        .initialize()
        .then((fulfilled) => {
          console.info('Data Source has been initialized!');
          return fulfilled;
        })
        .catch((err) => {
          console.error('Error during Data Source initialization', err);
          return null;
        });
    }
    return TypeOrm.instance;
  }
}

export default TypeOrm;
``` 
Then, at the **top level** of `hooks.server.ts` I start the Datasource and import `reflect-metadata`:
```typescript
import 'reflect-metadata';
...
await TypeOrm.getDb();
``` 
This appears to fire only once on startup, and rarely when hot reloading.

In my server load functions I then use the active record pattern:
```typescript
export const load: PageLoad = async () => {
  const test = new Test();
  test.name = 'test name';
  await test.save();
  const tests = await Test.find();
  const json = JSON.stringify(tests);
  return { json };
};
``` 
To use either entity manager or repository, bring in the singleton:
```typescript
  ...
  const db = await TypeOrm.getDb();
  const test = new Test();
  test.name = 'test name';
  db.manager.save(test)
``` 
However I'll note that the repository pattern doesn't appear to be creating new entities for me, but haven't put much effort into fixing it.

**Reply by saki82** [reply]
Hello @ngramp,
above solution works while running the dev server, thanks for sharing.
But how did you manage to build the project?
Since ```hooks.server.ts``` is not invoked during the vite build process (at least thats what im suspecting is the cause), the build fails:

```
vite v5.0.12 building SSR bundle for production...
✓ 998 modules transformed.

node:internal/event_target:1088
  process.nextTick(() => { throw err; });
                           ^
Error: DataSource is not set for this entity.
    at User.getRepository (/Users/athanassiosanastasiadis/Documents/Development/svelte-typeorm/node_modules/.pnpm/typeorm@0.3.19_mysql2@3.7.1_ts-node@10.9.2/node_modules/typeorm/repository/BaseEntity.js:78:19)
    at User.findOne (/Users/athanassiosanastasiadis/Documents/Development/svelte-typeorm/node_modules/.pnpm/typeorm@0.3.19_mysql2@3.7.1_ts-node@10.9.2/node_modules/typeorm/repository/BaseEntity.js:280:21)
    at file:///Users/athanassiosanastasiadis/Documents/Development/svelte-typeorm/.svelte-kit/output/server/entries/pages/items/_id_/_page.server.ts.js:5:25
    at ModuleJob.run (node:internal/modules/esm/module_job:218:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:329:24)
    at async Promise.all (index 4)
    at async analyse (file:///Users/athanassiosanastasiadis/Documents/Development/svelte-typeorm/node_modules/.pnpm/@sveltejs+kit@2.4.2_@sveltejs+vite-plugin-svelte@3.0.1_svelte@4.2.9_vite@5.0.12/node_modules/@sveltejs/kit/src/core/postbuild/analyse.js:66:16)
    at async MessagePort.<anonymous> (file:///Users/athanassiosanastasiadis/Documents/Development/svelte-typeorm/node_modules/.pnpm/@sveltejs+kit@2.4.2_@sveltejs+vite-plugin-svelte@3.0.1_svelte@4.2.9_vite@5.0.12/node_modules/@sveltejs/kit/src/utils/fork.js:22:16)
Emitted 'error' event on Worker instance at:
    at [kOnErrorMessage] (node:internal/worker:326:10)
    at [kOnMessage] (node:internal/worker:337:37)
    at MessagePort.<anonymous> (node:internal/worker:232:57)
    at [nodejs.internal.kHybridDispatch] (node:internal/event_target:814:20)
    at exports.emitMessage (node:internal/per_context/messageport:23:28)

Node.js v20.10.0
 ELIFECYCLE  Command failed with exit code 1.
```

It's the same error as if there's no ```hooks.server.ts``` present while running the dev server.
I tried invoking a vite plugin in ```vite.config.ts```, which calls ```TypeOrm.getDb()```, but unfortunately i get the same error:

```
// vite-plugin-svelte-typeorm.ts

import TypeOrm from './src/lib/server/data-source';

export default function vitePluginSvelteTypeOrm(options = {}) {
	return {
		name: 'vitePluginSvelteTypeOrm',
		async buildStart() {
			await TypeOrm.getDb();
		},
	};
}
```

```
// vite.config.ts

import { sveltekit } from '@sveltejs/kit/vite';
import { defineConfig } from 'vitest/config';
import vitePluginSvelteTypeOrm from './vite-plugin-svelte-typeorm';

export default defineConfig({
	ssr: {
		external: ['reflect-metadata'],
	},
	plugins: [sveltekit(), vitePluginSvelteTypeOrm()],
	test: {
		include: ['src/**/*.{test,spec}.{js,ts}'],
	},
	resolve: {
		alias: { $lib: './src/lib' },
	},
});
```
Thanks in advance,
Sakis

**Reply by ngramp** [reply]
Been a while, but iirc 'DataSource is not set for this entity' suggests TypeOrm loaded fine but it can't find the User entity.  Is it in the entities array? Is the db synced?

**Reply by saki82** [reply]
Yes it's on the entities array and db is synced. 

I can pinpoint the issue to hooks.server.ts not being called during the build, which leads to Typeorm.getDb() not being invoked and thus, new Datasource() not being instantiated. 
 
My workaround is to invoke Typeorm.getDb() in every single page.server.ts. This way the build succeeds, but its certainly not the most elegant. 😕

**Reply by Ennoriel** [reply]
Are you trying to generate a static site with the adapter static and `export const prerender = true`? If so, it should call `hooks.server.ts` on every request (you may need to provide an empty `handle` function):

```js
export async function handle({ event, resolve }) {
    return await resolve(event);
}
```

**Reply by saki82** [reply]
Hi @Ennoriel,
i am building with the node adapter in default ssr mode.
This is what my ```hooks.server.ts``` looks like:
```
export const handle = {
	auth: SvelteKitAuth({
		providers: [GitHub({ clientId: env.GITHUB_ID, clientSecret: env.GITHUB_SECRET })],
	}),
	db: TypeOrm.getDb(), <-- this is the db instantiation function
};
```
I suspect it is never called during build with the default ssr option

**Reply by Ennoriel** [reply]
I would try something like that:

```
const db = TypeOrm.getDb()

export const handle = {
	auth: SvelteKitAuth({
		providers: [GitHub({ clientId: env.GITHUB_ID, clientSecret: env.GITHUB_SECRET })],
	}),
	db,
};
```

in the build step, `handle` is anly called in static site generation when rendering each single route
